Use smart constructors in the parser

jstolarek opened the pull request:
This is the first chunk of work on introducing smart constructors (#421). It fixes #442 and otherwise simplifies or refactors multiple productions in the parser. I introduced a new module to store smart constructors. Many of the already existing functions have been moved from parser to that module.

One particular thing I would particularly like to hear feedback on are the naming conventions for the constructors. Names of most of the functions I introduced begin with `make_`. Some of the already existing functions follow that convention - c.f. `make_untyped_handler` in sugartypes.ml or `make_tuple` - but others don't (e.g. `datatype`, `cp_unit`). I am thinking of dropping `make_` prefix altogether to make the code more concise. Do you think this would be a good idea?

jamescheney commented:
I wonder if it would make sense to remove the "make_" prefix, but import the smartConstructors module as a qualified module name such as `Make` or `Mk`, so that you would write `Make.untyped_handler" or "Mk.untyped_handler" - using the module's namespace to decrease the clutter in the identifiers?  Or is that an antipattern?

dhil commented:
The more modules the merrier. I think it is a good idea to collect those utility functions in a module named `Make` or something descriptive. Organising code into to separate modules is generally a good idea.

dhil commented:
You can embed the module `SugarConstructors` directly into `SugarTypes`, e.g.
```ocaml
module Make: sig
   val funlit : ppos -> declared_linearity -> pattern list list -> block_body with_pos -> phrase
end = struct
  let funlit ....
end
```
That way the constructors would be available whenever `SugarTypes` are in scope.

slindley commented:
What @jamescheney suggests is idiomatic OCaml - not an antipattern

jstolarek commented:
One more question. Functions in `Make` module currently use `ppos`, i.e. positions produced by the parser, which differ from positions used in Sugartypes. Since constructors module is currently used only by the parser it is not a problem. But it will become a problem once I want to apply smart constructors in desugar modules. What would be a canonical way of solving such a problem in OCaml? Parametrizing `Make` module with a type of positions?

dhil commented:
> Functions in Make module currently use ppos, i.e. positions produced by the parser, which differ from positions used in Sugartypes.

It may be that I don't remember the details, but each (term) node in AST is paired with a position produced by the parser. The positions produced by the parser have type `SourceCode.pos`. Each desugaring pass operate over the AST from `Sugartypes`, so they ought to use `SourceCode.pos` too. 

The type checker uses both `SourceCode.pos` and `SourceCode.position`, where the latter is the type of "resolved" positions. Resolved positions are used in error messages.

jstolarek commented:
> The positions produced by the parser have type `SourceCode.pos`

Before switch to Menhir that was indeed the case. We just had this one `pos` function in the parser that produced a position of type `SourceCode.pos` for a whole production:

```
let pos () : Sugartypes.position = Parsing.symbol_start_pos (), Parsing.symbol_end_pos (), None
```

With Menhir that function no longer exists and we have access built-in Menhir functions that produce locations both for a whole production and its sub-productions. The type of positions produced by Menhir is `SourceCode.lexpos * SourceCode.lexpos`. So after a switch to Menhir I just created convenient wrappers around `sugartypes.with_pos`:

```
(* Convert position produced by a parser to Sugartypes position *)
let pos (start_pos, end_pos)= (start_pos, end_pos, None)

(* Wrapper around Sugartypes.with_pos.  Accepts parser positions. *)
let with_pos p = Sugartypes.with_pos (pos p)
``` 

As long as these functions were limited in scope to the parser this design worked nicely.
But now when they are in `Make` module that will get used in other places this no longer looks like a good design. Even now you will notice there were several places in sugarConstructors.ml where I had to explicitly call `Sugartypes.with_pos` and not the `Make.with_pos`.

jstolarek commented:
I attempted to solve the problem outlined above by using a functor, but I've been defeated by OCaml (or rather my lack of knowledge of OCaml). My work is on `js-T421-smart-constructors-functor` branch. Here's what I attempted to do. Firstly, abstract positions using a module:

```
module type Pos = sig
  type t
  val pos      : t -> Sugartypes.position
  val with_pos : t -> 'a -> 'a Sugartypes.with_pos
end
```

In the parser this module will implement simple wrappers around definitions in sugartypes.ml:


```
module ParserPosition : Pos = struct
  type t = SourceCode.lexpos * SourceCode.lexpos
  let pos (start_pos, end_pos) = (start_pos, end_pos, None)
  let with_pos p = Sugartypes.with_pos (pos p)
end
```

In all other modules `pos` will be an identity and `with_pos` will be an alias for `Sugartypes.with_pos`.  All the smart constructors will be defined in a functor that takes module of type `Pos`:

```
module SmartConstructors (Position : Pos)
       : (SmartConstructorsSig with type t := Position.t) = struct

  let pos      = Position.pos
  let with_pos = Position.with_pos

  let variable_pat ppos name =
    with_pos ppos (`Variable (make_untyped_binder name))
  
  (...)
```

with the signature `SmartConstructorsSig` being defined as:

```
module type SmartConstructorsSig = sig

  (* Positions *)
  type t
  val pos      : t -> Sugartypes.position
  val with_pos : t -> 'a -> 'a with_pos

  val variable_pat : t -> name with_pos -> pattern
  (...)
end
```

However, if I say this in the parser:

```
module ParserConstructors  = SmartConstructors(ParserPosition)
open ParserConstructors
```
or even
```
module ParserConstructors
       : (SmartConstructorsSig with type t := ParserPosition.t) =
  SmartConstructors(ParserPosition)
open ParserConstructors
```
then I get a compilation error on a `$loc` argument in a `with_pos $loc ...` expression:

```
File "parser.mly", line 682, characters 74-78:
Error: This expression has type Lexing.position * Lexing.position
       but an expression was expected of type ParserPosition.t
```
Except that the two types are equal. I was hoping to expose that equality with a destructive substitution `with type t := ParserPosition.t` but clearly that's not how it's supposed to be done. So, my questions are:

1. Is this even a good approach to solving the problem?
2. If so, how do I make it work?

I would also like to get rid of `pos` and `with_pos` from `SmartConstructorsSig` - these are only aliases - and use definitions from `Pos` directly. I think I know how to do it but first I need to get the code to compile :-)

rudihorn commented:
Try replacing the code with:

```
module ParserPosition : Pos with type t = Lexing.Position * Lexing.Position = struct
...

end

```


By saying the module is of type `Pos`, you are abstracting away the information about what `t` is, and the rest of the code will treat `t` as completely different to its underlying type.

jstolarek commented:
That works, thanks! This is strange though. I could swear I tried this before and got an error message complaining about mismatched signatures due to `t` being undefined when I tried to instantiate `SmartConstructors(ParserPosition)`. I guess some other fragment of the code must've been different as well.

rudihorn commented:
You probably in that case used destructive substitution rather than type equality, which gets rid of the type t completely. This means that the module is not of module type `Pos` anymore.

jstolarek commented:
Yes, I believe this is what I did. Thanks for explanation.

jstolarek commented:
I updated the PR to include this functor refactoring. One more question is whether it is possible to avoid repeating the long `SmartConstructorsSig` signature in both the `ml` and `mli` files?

A cool thing about this design is that we can now define:

```
module DummyPosition : Pos with type t = unit = struct
  type t          = unit
  let pos ()      = Sugartypes.dummy_position
  let with_pos () = Sugartypes.with_pos Sugartypes.dummy_position
end

module DummyMake = SmartConstructors(DummyPosition)
```

And use this module when constructing values in `desugar*`modules. Or perhaps even make this the default implementation, since we mostly construct things with dummy values (see #423).

jstolarek commented:
I cleaned up the patch using Rudi's suggestions and added documentation. If there are no change requests then I would like to merge it before moving to refactor desugar* modules.

dhil commented on core/parser.mly:
```diff
@@ -59,7 +60,9 @@ module ParserPosition
 end
 
-module ParserConstructors = SmartConstructors(ParserPosition)
+(* Instatntiate SugarConstructors functor and open it to avoid explicit name
```
s/Instatntiate/Instantiate

dhil replied:
s/Instatntiate/Instantiate

dhil commented on core/sugarConstructors.mli:
```diff
@@ -1,16 +1,13 @@
-module type SmartConstructorsSig = SmartConstructorsIntf.SmartConstructorsSig
+(* Actual implementation of smart constructors as a functor on a Pos module *)
+module SugarConstructors (Position : Pos)
+       : (SugarConstructorsSig with type t := Position.t)
```
Comment: You should be able to drop the parentheses in `(SugarConstructorsSig with type t := Position.t)`. Personally, I don't care whether we have a few extraneous parentheses.

jstolarek replied:
Yes, these are redundant. I don't know which style is considered better. Personally I found having these parens more readable.

dhil commented on core/sugarConstructors.ml:
```diff
@@ -2,18 +2,15 @@ open Sugartypes
-end
-
-module type SmartConstructorsSig = SmartConstructorsIntf.SmartConstructorsSig
+(* Import module signatures *)
```
Missing punctuation.

jstolarek replied:
I tend not to short (single-sentence) comments with punctuation - that's why it's consistently missing. Will fix.

dhil commented on core/sugarConstructors.ml:
```diff
@@ -2,18 +2,15 @@ open Sugartypes
 
-module SmartConstructors (Position : Pos)
-       : (SmartConstructorsSig with type t := Position.t) = struct
+(* Actual implementation of smart constructors as a functor on a Pos module *)
```
Missing punctuation.

dhil commented on core/sugarConstructors.ml:
```diff
@@ -2,18 +2,15 @@ open Sugartypes
+       : (SugarConstructorsSig with type t := Position.t) = struct
 
-  (** Positions *)
+  (** Convenient aliases for functions operating on positions *)
```
Missing punctuation.

dhil commented on core/sugarConstructors.ml:
```diff
@@ -227,6 +224,7 @@ module SmartConstructors (Position : Pos)
 
 end
 
+(* A default type of positions *)
```
Missing punctuation.

dhil commented on core/sugarConstructors.mli:
```diff
@@ -1,16 +1,13 @@
-  val pos      : t -> Sugartypes.position
-  val with_pos : t -> 'a -> 'a Sugartypes.with_pos
-end
+(* Import module signatures *)
```
Missing punctuation.

dhil commented on core/sugarConstructors.mli:
```diff
@@ -1,16 +1,13 @@
-       : (SmartConstructorsSig with type t := Position.t)
-
-module Make : (SmartConstructorsSig
+(* Modules for making nodes using various types of positions *)
```
Missing punctuation.

dhil commented on core/sugarConstructorsIntf.ml:
```diff
@@ -1,9 +1,32 @@
+   a way of converting such positions to Sugartypes.position and this is what
+   modules implementing this signature provide. *)
+module type Pos = sig
+  (* Type of positions *)
```
Missing punctuation.

dhil commented on core/sugarConstructorsIntf.ml:
```diff
@@ -1,9 +1,32 @@
+module type Pos = sig
+  (* Type of positions *)
+  type t
+  (* Convert a position to Sugartypes.position *)
```
Missing punctuation.

dhil commented on core/sugarConstructorsIntf.ml:
```diff
@@ -1,9 +1,32 @@
+  type t
+  (* Convert a position to Sugartypes.position *)
+  val pos      : t -> Sugartypes.position
+  (* Produce a syntax tree node with a position attached *)
```
Missing punctuation.

dhil commented on core/sugarConstructorsIntf.ml:
```diff
@@ -1,9 +1,32 @@
+(* An abstract type of positions and operations on them.  The core type of
+   positions used in the compilation pipeline is Sugartypes.position, which
+   again is an alias to SourceCode.pos.  However, in several places we operate
+   on positions of a different type.  Most importantly, the parser produces a
```
s/the parser produces a positions/the parser produces positions.

jstolarek commented:
@dhil, thanks! All done, except for parens around type signatures in modules.

jstolarek commented:
Merged as f06ca1ad
